### ISSUE TITLE
Reassigning `let` variable to subtype does not trigger refinement

### ISSUE THREAD

**Issue by jtbandes** [microsoft/TypeScript#54594]
# Bug Report

### 🔎 Search Terms

refinement instanceof assign

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about refinement

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.3#code/MYGwhgzhAEDC0G9oA8D8AuRZ0DsCuAtgEYCmATgL7QUBQokMAItCcgC4k4AmM8SamBNnzFyAGiLoIbMgEscAcyq0aIEm2jBoAXmg4SAdzgAKAJQBuGgHor0AHqoa0Z1t36jjM5Zv3HdAHRo-kSWNEA)

### 💻 Code

```ts
class C { x?: {a:number} }
class D extends C { x?: {a:number,b:string} }

let c = new C();
c = new D();

c.x?.b;  // error :(
```

### 🙁 Actual behavior

The type of `c` is still `C`

### 🙂 Expected behavior

After `c = new D()`, the type of `c` should be refined to `D`

**Comment by jcalz** [comment]
Duplicate of #46489, probably others too

Assignment narrowing only filters unions.
